File: sdks/rust/src/seed/resources/witness.rs
//! Witness-chain resource — `/api/v1/witness/*`.

use crate::error::Error;
use crate::seed::client::SeedClient;
use crate::seed::config::CallOptions;
use crate::seed::models::WitnessChain;

/// Witness endpoints.
pub struct WitnessResource<'c> {
    pub(crate) client: &'c SeedClient,
}

impl<'c> WitnessResource<'c> {
    /// `GET /api/v1/witness/chain` (allowlisted read).
    pub async fn chain(&self) -> Result<WitnessChain, Error> {
        self.client.request_get("/witness/chain").await
    }

    /// [`Self::chain`] with per-call [`CallOptions`] overrides.
    pub async fn chain_with(&self, opts: CallOptions) -> Result<WitnessChain, Error> {
        self.client.request_get_opts("/witness/chain", &opts).await
    }
}

File: sdks/node/src/seed/resources/witness.ts
/** GET /api/v1/witness/chain — Phase 1 resource. */

import type { CallOptions } from "../callOptions.js";

export interface WitnessEntry extends Record<string, unknown> {
  epoch: number;
  action?: string;
  signature?: string;
  timestamp?: string;
}

export interface WitnessChain extends Record<string, unknown> {
  entries?: WitnessEntry[];
  length?: number;
  head?: string;
}

type RequestFn = <T>(
  method: string,
  path: string,
  opts?: CallOptions & { idempotent?: boolean },
) => Promise<T>;

export interface WitnessResource {
  /** GET /api/v1/witness/chain — WiFi-read allowlist. */
  chain(opts?: CallOptions): Promise<WitnessChain>;
}

export function makeWitnessResource(request: RequestFn): WitnessResource {
  return {
    chain: (opts) =>
      request<WitnessChain>("GET", "/api/v1/witness/chain", {
        idempotent: true,
        ...(opts ?? {}),
      }),
  };
}
